From 817815adedee1de6aa98a739432c893db8852f52 Mon Sep 17 00:00:00 2001
From: "hanghang.luo" <hanghang.luo@amlogic.com>
Date: Fri, 12 May 2023 02:56:54 +0000
Subject: [PATCH 1/3] amlv4l2dec: CB1 drop v4l2 eos event [1/1]

PD#SWPL-119356

Problem:
first dq eos event, then the last frame will be dropped

Solution:
drop v4l2 eos event, dq last frame, next dq last empty buff,then flow
over

Verify:
Yocto

Change-Id: Ib1cdc70baaa1a8e550c81b97c24abf648da13696
Signed-off-by: hanghang.luo <hanghang.luo@amlogic.com>
---
 src/gstamlv4l2bufferpool.c | 19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

diff --git a/src/gstamlv4l2bufferpool.c b/src/gstamlv4l2bufferpool.c
index 7f7fec6..dcd6c7d 100644
--- a/src/gstamlv4l2bufferpool.c
+++ b/src/gstamlv4l2bufferpool.c
@@ -1400,6 +1400,11 @@ gst_aml_v4l2_buffer_pool_dqbuf(GstAmlV4l2BufferPool *pool, GstBuffer **buffer,
         goto eos;
     if (res != GST_FLOW_OK)
         goto dqbuf_failed;
+    if ( (group->buffer.flags & V4L2_BUF_FLAG_LAST) &&(group->buffer.bytesused == 0) )
+    {
+        GST_LOG_OBJECT (pool,"dequeued empty buffer");
+        return GST_AML_V4L2_FLOW_LAST_BUFFER;
+    }
 
     /* get our GstBuffer with that index from the pool, if the buffer was
      * outstanding we have a serious problem.
@@ -1568,9 +1573,19 @@ gst_aml_v4l2_buffer_pool_dequeue(GstAmlV4l2BufferPool *pool, GstBuffer **buffer,
     {
         GstFlowReturn res_event = gst_aml_v4l2_buffer_pool_dqevent(pool);
         if (res_event != GST_FLOW_OK)
-            return res_event;
+        {
+            /* when drive V4l2 receive cmd_stop, it will finish current decoding frame, then creat
+             * a EOS event and a empty buff. if gstreamer dq EOS event first ,the last frame will be drop,
+             * this a question
+             */
+            if (res_event == GST_AML_V4L2_FLOW_LAST_BUFFER)
+            {
+                GST_DEBUG_OBJECT(pool," reiceive EOS event, drop it");
+            }
+            else
+                return res_event;
+        }
     }
-
     if (res == GST_FLOW_CUSTOM_SUCCESS)
     {
         GST_LOG_OBJECT(pool, "nothing to dequeue");
-- 
2.25.1

